### ISSUE TITLE
getgrnam_r buffer too small in exec.c

### ISSUE THREAD

**Issue by EvanKuhn** [collectd/collectd#2208]
*   Version of collectd: 5.4.0
*   Operating system / distribution: Ubuntu 12.04.5 LTS

## Expected behavior

Exec plugins run as expected.

## Actual behavior

We've recently noticed that collectd no longer runs Exec plugins.  We are getting the following errors in /var/log/syslog:

```
collectd[31612]: exec plugin: Failed to get group information for group ``nagios'': Numerical result out of range
collectd[31612]: exec plugin: Failed to get group information for group ``nogroup'': Numerical result out of range
```

The Execs we have are configured to use the `nagios` and `nogroup` groups.

I tracked the error message down to a call to `getgrnam_r` in src/exec.c:409, just before it tries to fork the child process.  I was able to write a small program to reproduce the error on our system.  `getgrname_r` returns `ERANGE`, indicating that the `nambuf` buffer (2048 bytes) is too small.  We suspect that recent user and group additions on our systems require a larger buffer.

I further notice that every minute, collectd will open (and fail to close) 6 additional pipes per Exec plugin.  This eventually leads to the service running out of open file handles and becoming unresponsive, requiring a `kill -9` to terminate it.  Could this be related to the 3 pipes that `fork_child()` creates?

To fix this, I think we can either:
- simply increase the size of `nambuf` from 2kb to, say, 8kb
- if `getgrnam_r` fails and returns `ERANGE`, retry with a larger buffer, up to some max

## Steps to reproduce

*   I currently only have a test program that reproduces the error (below).
*   Or on a test system, add a large number of users and groups, or compile and run collectd with a smaller getgrnam_r buffer size.  I haven't tried these yet.

## Test program

```C
// Test getgrnam_r() failures due to a buffer that's too small.
// gcc groupinfo.c -o groupinfo && ./groupinfo

#include <sys/types.h>
#include <grp.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>

int main(int argc, char** argv) {
  if (argc < 3) {
    printf("usage: ./groupinfo <group> <buflen>\n");
    return 1;
  }

  int buflen = atoi(argv[2]);
  if (buflen == 0) {
    printf("Unable to parse '%s' to an integer buffer length\n", argv[2]);
    return 1;
  }

  char* groupname = argv[1];
  struct group *grp = NULL;
  struct group gr;
  char* nambuf = malloc(buflen); // Smallest size that works is 2392
  char errbuf[2048];

  // Make sure getgrnam_r is supported
#if  _POSIX_C_SOURCE >= 1 || _XOPEN_SOURCE || _BSD_SOURCE || _SVID_SOURCE || _POSIX_SOURCE
  printf("getgrnam_r supported\n");
#else
  printf("getgrnam_r NOT supported\n");
  return 1;
#endif

  printf("Getting info for group '%s' with buflen=%d\n", groupname, buflen);
  int status = getgrnam_r(groupname, &gr, nambuf, buflen, &grp);

  if (0 != status) {
    if (status == ERANGE) printf("Got ERANGE\n");
    strerror_r(errno, errbuf, sizeof(errbuf));
    printf("getgrnam_r failed for group '%s': %s\n", groupname, errbuf);
  }
  else {
    printf("Got group info:\n");
    printf("The group name is: '%s'\n", grp->gr_name);
    printf("The password is:   '%s'\n", grp->gr_passwd);
    printf("The gid is:        '%u'\n", grp->gr_gid);
  }
  return 0;
}
```

Sample output:

```
$ ./groupinfo ekuhn 2048
getgrnam_r supported
Getting info for group 'ekuhn' with buflen=2048
Got ERANGE
getgrnam_r failed for group 'ekuhn': Numerical result out of range

$ ./groupinfo ekuhn 2500
getgrnam_r supported
Getting info for group 'ekuhn' with buflen=2500
Got group info:
The group name is: 'ekuhn'
The password is:   'x'
The gid is:        '1609'
```

**Comment by rubenk** [comment]
Probably best to dynamically size the buffer based on `sysconf(_SC_GETGR_R_SIZE_MAX)`.

**Comment by mwtzzz** [comment]
Any progress on this? I just noticing the same issue on a cluster of graphite servers that I manage, in an environment where we have recently made changes to our unix groups,

**Comment by rubenk** [comment]
Should be fixed by 24bdf52decaf93359aab0364fb11420af5cd4a21
